### PR TITLE
Do not persist winforms data to disk

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
There is a possible race condition in this cde (originally documented) that could lead us to persist data to disk that does not match what the host thinks the state of things are.  i.e. we compute something is a form, then remote that information over to VS.  THe processing of that data on the VS side happens at some point in the future, but we still persist what we think VS knows at that point.

If this gets out of sync (i.e. because a message is lost) user might be in a broken state until they can do something that forces reprocessing.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/VisualStudioDesignerAttributeService.cs [hunk 1]
```diff
@@ -108,7 +108,7 @@ private async Task StartAsync(CancellationToken cancellationToken)
         private async Task StartWorkerAsync(CancellationToken cancellationToken)
         {
             _workQueue = new AsyncBatchingWorkQueue<DesignerAttributeData>(
-                TimeSpan.FromSeconds(1),
+                TimeSpan.FromMilliseconds(500),
```
Comment (jmarolf) [inline src/VisualStudio/Core/Def/Implementation/DesignerAttribute/VisualStudioDesignerAttributeService.cs:111]: what is the rationale for this change?

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/DesignerAttribute/VisualStudioDesignerAttributeService.cs:111]: i actually reduce it even more in the next PR.  the point of these delays is to allow us to hear about a large set of result from oop at once, and then report them at once to VS in one go.  i.e. we don't want to hear about 100 files, and then send 100 calls to update VS (each which might hit the UI thread).  Instead, it's better to hear about 100 files, then report it in one go to VS on the UI thread once.

So the goal is simply to give enough time to hear about a lot, but not so long that it feels slow to the user for thigns to update.  250ms seems to work well.  It allows for hearing about tons of files, not saturating the UI thread, and providing low latency between hearing about changes and updating the UI with them.  Note: my next PR adds a comment on this.

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs [hunk 2]
```diff
@@ -74,121 +90,82 @@ private async Task AnalyzeProjectAsync(Project project, Document? specificDocume
...
-            //
-            //  !  is safe here as `i.changed` implies `i.info` is non-null.
-            var changedInfos = latestInfos.Where(i => i.changed).Select(i => i.data!.Value).ToList();
-            if (changedInfos.Count > 0)
+            var latestData = await ComputeLatestDataAsync(
+                project, specificDocument, projectVersion, cancellationToken).ConfigureAwait(false);
+
+            var changedData =
+                latestData.Where(d => !_documentToLastReportedInformation.TryGetValue(d.document.Id, out var existingInfo) || existingInfo.category != d.data.Category)
+                          .ToImmutableArray();
```
Comment (jasonmalinowski) [inline src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:101]: If the user had a file that we classified, and then edits the code such that ComputeLatestDataAsync no longer returns anything for that file, do we need to report the now lack of data? And clear out the dictionary in that case?

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs [hunk 3]
```diff
@@ -74,121 +90,82 @@ private async Task AnalyzeProjectAsync(Project project, Document? specificDocume
...
+            using var _ = ArrayBuilder<Task<(Document document, DesignerAttributeData data)>>.GetInstance(out var tasks);
             foreach (var document in project.Documents)
             {
                 // If we're only analyzing a specific document, then skip the rest.
                 if (specificDocument != null && document != specificDocument)
                     continue;
 
-                tasks.Add(ComputeDesignerAttributeDataAsync(
-                    storage, projectVersion, designerCategoryType, document, cancellationToken));
+                // If we don't have a path for this document, we cant proceed with it.
```
Comment (jasonmalinowski) [inline src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:129]:
```suggestion
                // If we don't have a path for this document, we can't proceed with it.
```

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs [hunk 4]
```diff
@@ -74,121 +90,82 @@ private async Task AnalyzeProjectAsync(Project project, Document? specificDocume
...
+                // We need that path to inform the project system which file we're referring to.
+                if (document.FilePath == null)
+                    continue;
+
+                // If nothing has changed at the top level between the last time we analyzed this document and now, then
+                // no need to analyze again.
+                if (_documentToLastReportedInformation.TryGetValue(document.Id, out var existingInfo) &&
+                    existingInfo.projectVersion == projectVersion)
```
Comment (jasonmalinowski) [inline src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:137]: What's the threading guarantees here? Can RemoveDocumentAsync earlier in the file be running at the same time as this? Is there some risk that we'd remove the document, then process it here, and add it back?

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs [hunk 5]
```diff
@@ -74,121 +90,82 @@ private async Task AnalyzeProjectAsync(Project project, Document? specificDocume
...
-                    // We were able to read out the old data, and it matches our current project
-                    // version.  Just return back that nothing changed here.  We won't tell VS about
-                    // this, and we won't re-persist this later.
-                    return default;
-                }
+                Contract.ThrowIfNull(document.FilePath);
 
                 // We either haven't computed the designer info, or our data was out of date.  We need
                 // So recompute here.  Figure out what the current category is, and if that's different
                 // from what we previously stored.
```
Comment (jasonmalinowski) [inline src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:157]: Comment seems stale -- should be deleted now?

**Comment by CyrusNajmabadi** [comment]
> Looks OK and I'll never complain about the removal of the a bunch of storage stuff. A bit worried about the use of ConcurrentDictionary in a place or two where if various members here can be running in parallel strange things might happen.

So a core benefit of IncrementalAnalyzer is that it makes its requests to us serially.  I do think i could just make this a Dictionary.  however, then i have to be very careful about my own reads/writes in the tasks i spin off to process all docs in a project in parallel.

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs [hunk 6]
```diff
@@ -28,20 +28,36 @@ internal sealed partial class RemoteDesignerAttributeIncrementalAnalyzer : Incre
...
 
-        public RemoteDesignerAttributeIncrementalAnalyzer(Workspace workspace, RemoteEndPoint endPoint)
+        public RemoteDesignerAttributeIncrementalAnalyzer(RemoteEndPoint endPoint)
         {
             _endPoint = endPoint;
-            _storageService = workspace.Services.GetRequiredService<IPersistentStorageService>();
         }
 
-        public override Task RemoveProjectAsync(ProjectId projectId, CancellationToken cancellationToken)
+        public override async Task RemoveProjectAsync(ProjectId projectId, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:43]: @mavasani for information here.  IN the case of solution-unload, how to incremetnal analyzers hear about the 'removal'?  Thanks!

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:101]: we do report the *lack* of data.  specifically, we report changes from whatever we last reported.  We don't actually clear out the dictionary because we want to store in teh dictionary the version of the last report, taht way we can ignore future events (that don't chanage hte state).  If we remove things from teh dictionary, that is akin to saying "we've never reported anything", and then we will just keep notifying VS of "it's not designable, it's not designable, it's not designable".

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:137]: it cannot be running at the same time.  all the incremental-analyzer entrypoints are serialized.  the only concurrency is from me spitting out tons of tasks (that i when-all) in one of those incremental-analyzer entrypoints.

@mavasani to confirm.

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:157]: it's still accurate.  we have to comptue this and compare against whatever we last reported.  teh difference now is that we keep track of what we reported in memory instead of on disk.

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:43]: Current implementation invokes RemoveDocumentAsync for each document in the solution: http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.cs,407. It seems it can certainly do batching here and just invoke RemoveProjectAsync for all projects in the solution.

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 4 above)
Comment (mavasani) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:137]: > all the incremental-analyzer entrypoints are serialized

Yes, that is correct

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:43]: on interesting.  so there's no real point in responding to RemoveProject if i'm also responding to RemoveDocument, correct?

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/VisualStudioDesignerAttributeService.cs (hunk 1 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/DesignerAttribute/VisualStudioDesignerAttributeService.cs:111]: "250ms seems to work well."

The PR is changing it to 500ms though?

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:101]: Ah got it; all documents being queried still get added to the dictionary no matter what. Makes sense.

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 5 above)
Comment (jasonmalinowski) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:157]: Well, it's still correct, but previously that function was doing the storage comparison. Now this function is blissfully ignorant of how it's being used -- does the comment move elsewhere then?

**Comment by jasonmalinowski** [comment]
@CyrusNajmabadi Sounds good.

File: src/VisualStudio/Core/Def/Implementation/DesignerAttribute/VisualStudioDesignerAttributeService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/DesignerAttribute/VisualStudioDesignerAttributeService.cs:111]: next pr reduces it further :)

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:157]: i'll revise in next pr.

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:157]: revised in next pr.

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:43]: We invoke RemoveProject for enqueued project items for which `solution.GetProject(id)` returns null, i.e. project has been removed and we cannot find the documents for it.

[the remark this answers is not in the document]
Comment (mavasani) [reply src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:43]: http://sourceroslyn.io/#Microsoft.CodeAnalysis.Features/SolutionCrawler/WorkCoordinator.LowPriorityProcessor.cs,153

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs [hunk 7]
```diff
@@ -28,20 +28,36 @@ internal sealed partial class RemoteDesignerAttributeIncrementalAnalyzer : Incre
         /// </summary>
         private readonly RemoteEndPoint _endPoint;
 
-        private readonly IPersistentStorageService _storageService;
+        /// <summary>
+        /// Keep track of the last information we reported.  We will avoid notifying the host if we recompute and these
```
Comment (tmat) [inline src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:32]: >host [](start = 89, length = 4)

Should this rather say "the client", not "the host"?

File: src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs [hunk 8]
```diff
@@ -74,121 +90,82 @@ private async Task AnalyzeProjectAsync(Project project, Document? specificDocume
...
-            // Note: we have a potential race condition here.  Specifically, for simplicity, the VS
-            // side will return immediately, without actually notifying the project system.  That
-            // means that we could persist the data to local storage that isn't in sync with what
-            // the project system knows about.  i.e. if VS is closed or crashes before that
-            // information is persisted, then these two systems will be in disagreement.  this is
-            // believed to not be a big issue given how small a time window this would be and how
-            // easy it would be to get out of that state (just edit the file).
-
-            await PersistLatestInfosAsync(project.Solution, projectVersion, latestInfos, cancellationToken).ConfigureAwait(false);
+            // Now, keep track of what we've reported to the host so we won't report unchanged files in the future.
```
Comment (tmat) [inline src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs:111]: >host [](start = 61, length = 4)

Client?
